src/server/conn_context.h

```diff
@@ -28,43 +27,47 @@ struct FlowInfo;
 // Used for storing MULTI/EXEC commands.
 class StoredCmd {
  public:
-  StoredCmd(const CommandId* cid, bool own_args, CmdArgList args);
+  // Deep copy of args, creates backing storage internally.
+  StoredCmd(const CommandId* cid, ArgSlice args, facade::ReplyMode mode = facade::ReplyMode::FULL);
 
-  // Create on top of already filled tightly-packed buffer.
-  StoredCmd(std::string&& buffer, const CommandId* cid, CmdArgList args, facade::ReplyMode mode);
+  // Shallow copy of args.
+  StoredCmd(const CommandId* cid, facade::ParsedArgs args)
+      : cid_{cid}, args_{args}, reply_mode_(facade::ReplyMode::FULL) {
+  }
 
   size_t NumArgs() const {
-    return std::visit(Overloaded{//
-                                 [](const OwnStorage& s) { return s.sizes.size(); },
-                                 [](const CmdArgList& s) { return s.size(); }},
-                      args_);
+    return args_.size();
+  }
+
+  size_t UsedMemory() const {
+    return backed_ ? backed_->HeapMemory() + sizeof(*backed_) : 0;
   }
 
-  size_t UsedMemory() const;
   facade::CmdArgList ArgList(CmdArgVec* scratch) const;
   std::string FirstArg() const;
 
-  const CommandId* Cid() const;
+  const CommandId* Cid() const {
+    return cid_;
+  }
 
-  facade::ReplyMode ReplyMode() const;
+  facade::ReplyMode ReplyMode() const {
+    return reply_mode_;
+  }
 
  private:
-  const CommandId* cid_;  // underlying command
-  struct OwnStorage {
-    std::string buffer;                   // underlying buffer
-    absl::FixedArray<uint32_t, 4> sizes;  // sizes of arg part
-    explicit OwnStorage(size_t sz) : sizes(sz) {
-    }
-  };
+  const CommandId* cid_;     // underlying command
+  facade::ParsedArgs args_;  // arguments
 
-  std::variant<OwnStorage, CmdArgList> args_;  // args storage
-  facade::ReplyMode reply_mode_;               // reply mode
+  // TODO: we could optimize the storage further by introducing StoredCmdCollection and
+  // keep the backing storage there. Then this class will only use shallow copies.
+  std::unique_ptr<cmn::BackedArguments> backed_;
```

Comment:
it seems like you can just steal the BackedArguments form the queue without much hesitation

Reply:
dispatch queue what in the future will be just ao object passed around